server/src/domain/access/access.core.ts

```diff
@@ -126,25 +145,42 @@ export class AccessCore {
       case Permission.ASSET_DOWNLOAD:
         return !!authUser.isAllowDownload && (await this.repository.asset.hasSharedLinkAccess(sharedLinkId, id));
 
-      case Permission.ASSET_UPLOAD:
-        return authUser.isAllowUpload;
-
       case Permission.ASSET_SHARE:
         // TODO: fix this to not use authUser.id for shared link access control
         return this.repository.asset.hasOwnerAccess(authUser.id, id);
 
-      case Permission.ALBUM_READ:
-        return this.repository.album.hasSharedLinkAccess(sharedLinkId, id);
-
-      case Permission.ALBUM_DOWNLOAD:
-        return !!authUser.isAllowDownload && (await this.repository.album.hasSharedLinkAccess(sharedLinkId, id));
-
       default:
         return false;
     }
   }
 
   private async checkAccessOther(authUser: AuthUserDto, permission: Permission, ids: Set<string>) {
+    switch (permission) {
+      case Permission.ALBUM_READ: {
+        const isOwner = await this.repository.album.checkOwnerAccess(authUser.id, ids);
+        const isShared = await this.repository.album.checkSharedAlbumAccess(authUser.id, setDifference(ids, isOwner));
```

Comment:
Did this still execute a query with zero records?

Reply:
All access repository methods I'm adding check for set size, and return early if the provided set is empty. I think encapsulating the logic there will simplify the code and reduce the need for each caller to validate beforehand.

Reply:
The case I am talking about is for a single album id. It will wait for is owned and then (, even if that is already true), it will wait for is shared to run. I am just mentioning that this is different than how it used to work and the number of calls for the album checks is now 2 sequential _always_, instead of one and _sometimes_ two calls.

Reply:
This is only applicable to cases that need one of many options to be fulfilled for it to be true. Maybe this is good enough for now. In the future I may look at combining these cases into a single database request instead of two

Reply:
If `ids` has a single id, and `checkOwnerAccess` validates that the id is accessible, then `isOwner` will also be a set with that single id.

In that scenario, `setDifference(ids, isOwner)` will return an empty set. `checkSharedAlbumAccess` will return early without actually making a second query, because of its first conditional `if (albumIds.size === 0) { return new Set(); }`. That should be matching the same behavior as we have today

Reply:
Ah, I misunderstood that by access checks you were referring to the repo calls as well. Yes, this works great!